homeassistant/helpers/entity_component.py

```diff
@@ -189,41 +271,64 @@ def __init__(self, component, scan_interval, entity_namespace):
+        ).result()
 
-            self.component.update_group()
+    @asyncio.coroutine
+    def async_add_entities(self, new_entities):
+        """Add entities for a single platform async.
 
-            if self._unsub_polling is not None or \
-               not any(entity.should_poll for entity
-                       in self.platform_entities):
-                return
+        This method must be run in the event loop.
+        """
+        for entity in new_entities:
+            ret = yield from self.component.async_add_entity(entity, self)
+            if ret:
+                self.platform_entities.append(entity)
 
-            self._unsub_polling = track_utc_time_change(
-                self.component.hass, self._update_entity_states,
-                second=range(0, 60, self.scan_interval))
+        self.component.async_update_group()
 
-    def reset(self):
-        """Remove all entities and reset data."""
+        if self._async_unsub_polling is not None or \
+           not any(entity.should_poll for entity
+                   in self.platform_entities):
+            return
+
+        self._async_unsub_polling = async_track_utc_time_change(
+            self.component.hass, self._update_entity_states,
+            second=range(0, 60, self.scan_interval))
+
+    def async_reset(self):
+        """Remove all entities and reset data.
+
+        This method must be run in the event loop.
+        """
         for entity in self.platform_entities:
-            entity.remove()
-        if self._unsub_polling is not None:
-            self._unsub_polling()
-            self._unsub_polling = None
+            entity.async_remove()
+
+        if self._async_unsub_polling is not None:
+            self._async_unsub_polling()
+            self._async_unsub_polling = None
 
+    @callback
     def _update_entity_states(self, now):
-        """Update the states of all the polling entities."""
-        with self.component.lock:
-            # We copy the entities because new entities might be detected
-            # during state update causing deadlocks.
-            entities = list(entity for entity in self.platform_entities
-                            if entity.should_poll)
+        """Update the states of all the polling entities.
+
+        This method must be run in the event loop.
+        """
+        tasks = []
+        # We copy the entities because new entities might be detected
+        # during state update causing deadlocks.
+        entities = list(entity for entity in self.platform_entities
+                        if entity.should_poll)
 
         for entity in entities:
-            entity.update_ha_state(True)
+            tasks.append(self.component.hass.loop.create_task(
+                entity.async_update_ha_state(True)
+            ))
+
+        # wait until all updates are done for protect loop for spaming
+        asyncio.gather(*tasks, loop=self.component.hass.loop)
```

Comment:
This does not prevent spamming. Since it's based on a callback, it will just be invoked again by the event listener.